### PR TITLE
Add Optimistic Sync Scenario Testing

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

End To End Testing

**What does this PR do? Why is it needed?**

This introduces optimistic sync testing in our e2e scenario test suite. In previous PRs, #10696 and #10808 we introduced the ability for our e2e test runner to perform scenario testing along with the ability to run an engine api proxy. We now combine these 2 feature with the ability to test optimistic sync.

- [x] Add in new scenario test for optimistic sync
- [x] Add in a few new features to our engine proxy, namely the ability to remove interceptors along with the ability to broadcast delayed queued up http requests.
- [x] Add in a few new e2e interfaces for more granular testing
- [x] Add in an optimistic sync evaluator.
  
**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

### CODE REVIEW

File: testing/middleware/engine-api-proxy/proxy.go [hunk 1]
```diff
@@ -111,8 +112,8 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		p.cfg.logger.WithError(err).Error("Could not parse request")
 		return
 	}
-	// Check if we need to intercept the request with a custom response.
-	hasIntercepted, err := p.interceptIfNeeded(requestBytes, w)
+	// Check if we need to intercept the request with a custom responseGen.
```
Comment (nisdas) [inline testing/middleware/engine-api-proxy/proxy.go:115]:
```suggestion
	// Check if we need to intercept the request with a custom response.
```

File: testing/middleware/engine-api-proxy/proxy.go [hunk 2]
```diff
@@ -124,9 +125,9 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	p.proxyRequest(requestBytes, w, r)
 }
 
-// AddRequestInterceptor for a desired json-rpc method by specifying a custom response
+// AddRequestInterceptor for a desired json-rpc method by specifying a custom responseGen
```
Comment (nisdas) [inline testing/middleware/engine-api-proxy/proxy.go:128]:
```suggestion
// AddRequestInterceptor for a desired json-rpc method by specifying a custom response
```

File: testing/middleware/engine-api-proxy/proxy.go [hunk 3]
```diff
@@ -136,9 +137,40 @@ func (p *Proxy) AddRequestInterceptor(rpcMethodName string, response interface{}
...
+		}
+		p.cfg.logger.Infof("Received response %s for backed up request for method %s", http.StatusText(res.StatusCode), rpcMethodName)
+	}
+	delete(p.backedUpRequests, rpcMethodName)
+}
+
 // Checks if there is a custom interceptor hook on the request, check if it can be
-// triggered, and then write the custom response to the writer.
-func (p *Proxy) interceptIfNeeded(requestBytes []byte, w http.ResponseWriter) (hasIntercepted bool, err error) {
+// triggered, and then write the custom responseGen to the writer.
```
Comment (nisdas) [inline testing/middleware/engine-api-proxy/proxy.go:172]:
```suggestion
// triggered, and then write the custom response to the writer.
```

**Review by rauljordan — APPROVED** [review]
This is really awesome. Great job on putting this together and great to see it all passing

**Pull request merged**